### PR TITLE
[WIP] ppx_deriving.4.4 release preparation branch

### PR DESCRIPTION
**Pull request by gasche**
* Restore support for OCaml 4.02.3
  ocaml-ppx/ppx_deriving#188
  (ELLIOTTCABLE)
* workaround Location.input_filename being empty
  when using reason-language-server
  ocaml-ppx/ppx_deriving#196
  (Ryan Artecona)
* Add support for OCaml 4.08.0
  ocaml-ppx/ppx_deriving#193, ocaml-ppx/ppx_deriving#197, ocaml-ppx/ppx_deriving#200
  (Gabriel Scherer)

### CODE REVIEW

**Comment by gasche** [comment]
The revdeps look good (there are few 4.08-available dependencies to test, but some like `lens` are and they seem to work), so I'll cut a real release and un-draft this PR.

**Comment by kit-ty-kate** [comment]
Once OCaml 4.08 enabled there was a lot of revdeps failures. I fixed most of them but this one seems rather weird:
```
#=== ERROR while compiling nocrypto.0.5.4-1 ===================================#
# context              2.0.4 | linux/x86_64 | ocaml-base-compiler.4.08.0 | git+file:///home/opam/opam-repository
# path                 ~/.opam/4.08/.opam-switch/build/nocrypto.0.5.4-1
# command              ~/.opam/4.08/bin/ocaml pkg/pkg.ml build --pinned false --tests false --jobs 1 --with-lwt true --xen false --freestanding false
# exit-code            1
# env-file             ~/.opam/log/nocrypto-17-c7ac5d.env
# output-file          ~/.opam/log/nocrypto-17-c7ac5d.out
### output ###
[...]
# + ocamlfind ocamlopt -pack -g -bin-annot -package ppx_sexp_conv -package bytes -package sexplib -package cstruct -I src src/native.cmx src/uncommon.cmx src/base64.cmx src/hash.cmx src/cipher_stream.cmx src/ccm.cmx src/numeric.cmx src/gcm.cmx src/cipher_block.cmx src/fortuna.cmx src/hmac_drgb.cmx src/rng.cmx src/rsa.cmx src/dsa.cmx src/dh.cmx -o src/nocrypto.cmx
# Fatal error: exception Invalid_argument("index out of bounds")
# Command exited with code 2.
```
There are a few broken packages to deal with but let's leave it like that for today

**Comment by mseri** [comment]
That's quite interesting, in principle it should not depend on `ppx_deriving` but just on `ppxlib`, right?

**Comment by gasche** [comment]
@kit-ty-kate thanks! I think that the failure that you are seeing on nocrypto is a 4.08.0 compiler bug that we have fixed in the 4.08 branch upstream, https://github.com/ocaml/ocaml/pull/8770

**Comment by gasche** [comment]
I studied more revdeps failures, and I didn't find anything very suspicious. (ppx_deriving_argparse is failing because of a hardcoded .cmxs path that changed when we moved to a Dune build, I can send a fix upstream and ask for a new release.) There are some unrelated failures and some plugins that really need a 4.08 upgrade.

I will complete the release and rebase this PR to point to the definitive archive. Thanks again @kit-ty-kate for the work triaging the issues and marking 4.08 incompatibility in dependent packages.

**Comment by gasche** [comment]
Ahem. I force-pushed to my github clone out of habit, and lost the extra commits that @kit-ty-kate had pushed. Apologies... @kit-ty-kate, can I find them in your own github fork? (I just looked but I don't see the branch.)

**Comment by gasche** [comment]
(No worries, I just got the missing patches from github: https://github.com/ocaml/opam-repository/compare/6506d59227c0380f861d6a647558cecb87673f55..8be19ff9ffa3517d2d749d13a847e589536bdfcc.patch )

**Comment by kit-ty-kate** [comment]
I had them locally. It was pushed on your branch directly. You might want to use `git push --force-with-lease` by default, it's very handy

**Comment by kit-ty-kate** [comment]
oops sorry my bad

**Comment by gasche** [comment]
Wow, `--force-with-lease` sounds very handy, thanks! (I wonder if there is a way to make it the default `--force` behavior.)

I just pushed an updated version of the branch that should be final (CI can always tell otherwise). I restricted the package to (< 4.09), so  that we avoid use-side build failures in the future. I restricted ppx_deriving_argparse to 4.3 instead of 4.4 (the failure is not new).

**Comment by kit-ty-kate** [comment]
> I restricted ppx_deriving_argparse to 4.3 instead of 4.4 (the failure is not new).

Are you sure? [check.ocamllabs.io](http://check.ocamllabs.io) tells me otherwise: http://check.ocamllabs.io/log/1562505242-7152ba482366e28003d2457509dd618295bc655d/4.07.1/good/ppx_deriving_argparse.0.0.4 (but that's not really important anyway)

**Comment by gasche** [comment]
The failure is only in the `make test` rule, but I thought I reproduced it locally under 4.3 and 4.07. Let me try again.

**Comment by gasche** [comment]
```
+ ocamlfind ocamlc -c -g -bin-annot -safe-string -ppx 'ocamlfind ppx_deriving/ppx_deriving src/ppx_deriving_argparse.cma /home/gasche/.opam/4.07.1/lib/ppx_deriving/ppx_deriving_show.cma' -I /home/gasche/.opam/4.07.1/lib/ppx_deriving/ -package result -package oUnit -w @5@8@10@11@12@14@23@24@26@29@40 -I src_test -I src -o src_test/test.cmo src_test/test.ml
File "_none_", line 1:
Error: Cannot load /home/gasche/.opam/4.07.1/lib/ppx_deriving/ppx_deriving_show.cmxs: cannot find file /home/gasche/.opam/4.07.1/lib/ppx_deriving/ppx_deriving_show.cmxs in search path
```

**Comment by kit-ty-kate** [comment]
thanks a lot!

**Comment by kit-ty-kate** [comment]
Ping to maintainers of packages of which latest version is not available with OCaml 4.08:
* @masashi-y for `ppx_deriving_argparse`
* @gasche for `ppx_deriving_protobuf` and `ppx_deriving_yojson`
* @samoht for `ppx_deriving_rpc`
* @fpottier for `visitors`

**Comment by fpottier** [comment]
Well, someone (@kate or @gasche?) marked `visitors` as incompatible with 4.08, without telling me. Would you care to explain what the problem is? Is it a problem with my code, or with some package that `visitors` depends upon, maybe `ppx_deriving` or `ppx_tools`?

**Comment by gasche** [comment]
That's part of the maintenance operations that we (most of that work is done by @kit-ty-kate) perform when the build of a package breaks (the idea is to mark it incompatible as soon as possible to avoid end-users seeing build failures). You were just "told" by the cc above: -- done by @kit-ty-kate when the change to the `visitors` description was merged in the main repository.

Here is what I get if I try to build `visitors`  on 4.08:

```
+ ocamlfind ocamlc -c -g -safe-string -package compiler-libs.common -package ppx_deriving.api -w A-4-44 -package result -o VisitorsCompatibility.cmo VisitorsCompatibility.ml
File "VisitorsCompatibility.cppo.ml", line 128, characters 6-21:
Error: The constructor Otag expects 2 argument(s),
       but is applied here to 3 argument(s)
```
